Add 96Boards WisTrio board support

Mani-Sadhasivam opened the pull request:
This PR adds 96Boards WisTrio board support with STM32L1 SoC series and HAL.

Fixes #9333

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
ext: hal: st: stm32cube: Add HAL for the STM32L1x series

This commit adds the STM32CUBE HAL for STM32L1x series MCUs.

Origin: ST Microelectronics
License: BSD-3-Clause
URL: http://www.st.com/en/embedded-software/stm32cubel1.html
Commit: 1.8.0
Purpose: HAL layer for STM32L1
Maintained-by: External

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: pinmux: Add STM32L1X pinmux support

Add pinmux support for STM32L1X SoC series.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: clock_control: Add STM32L1X clock support

Add clock support for STM32L1X SoC series.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: i2c: Add STM32L1X I2C support

Add I2C support for STM32L1X SoC series based on I2C_STM32_V1
driver.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
boards: arm: Add 96Boards WisTrio LoRa Tracker board

Add board support for 96Boards WisTrio LoRa Tracker board from
RAK Wireless. This board is one of the 96Boards IoT Edition platform.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

erwango commented:
@nashif, I'm surprised with the warning on Cube files license. It's usual BSD-3 clause as used in already in tree versions. Is there anything new on that front?

Mani-Sadhasivam commented:
@erwango Do you have any idea about the Sanity check failure?

erwango commented:
@Mani-Sadhasivam, please fix Licence issues for files that are not under ext/. For files under ext/, warning will be discarded.

Mani-Sadhasivam commented:
@erwango Fixed the license for non ext/ files. But I'm not sure about the shippable failure for the testcase.

erwango commented on soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb:
```diff
@@ -0,0 +1,28 @@
+
+config NUM_IRQS
+	int
+	default 32
```
Can you double check this value?
You're using IRQ num > 32 in dts, this should be causing the issue detected by shippable

Mani-Sadhasivam replied:
Ah. It is 45. Updated in next revision!

dbkinder reviewed (approved):
Doc additions LGTM, thanks!

erwango commented:
@Mani-Sadhasivam , can you rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/12024, this is GPIO driver refactoring, will spare you some code (bit late though ;-))

Mani-Sadhasivam commented:
@erwango I'd like to rebase once that PR gets merged. Or we can do the conversion as an incremental PR after merging this!

erwango commented:
@Mani-Sadhasivam, this is now merged.  Please take it into account as of now, I prefer not merging code I know it is already obsolete. Besides, due to factorization, this will be less code to review :-)

erwango commented:
@Mani-Sadhasivam, btw, and so you're not disappointed, there are significant changes on going (particularly on dt front), which might have impact on SoC porting and general and hence on this PR. These changes are meant to be integrated in 1.14 LTS.
I'm quite happy to see L1 port coming (thank you for that), but for coherency, I'd like STM32L1 port to be identical to other STM32 series, specially if merged in LTS. So I'll be more than happy to have this PR merged for 1.14, but this means that you'll have to integrate all these on going changes in a short period (and even though, I can't swear it'll will make it).

tagunil commented on soc/arm/st_stm32/stm32l1/Kconfig.series:
```diff
@@ -0,0 +1,18 @@
+config SOC_SERIES_STM32L1X
+	bool "STM32L1x Series MCU"
+	select CPU_CORTEX_M3
+	select CPU_CORTEX_M_HAS_VTOR
```
You don't need to select CPU_CORTEX_M_HAS_VTOR here because it's already selected by CPU_CORTEX_M3.

Mani-Sadhasivam replied:
Ack

nashif commented:
> @nashif, I'm surprised with the warning on Cube files license. It's usual BSD-3 clause as used in already in tree versions. Is there anything new on that front?

bsd-new is bsd-3 clause

Mani-Sadhasivam commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -43,6 +43,8 @@
 #elif defined(CONFIG_SOC_SERIES_STM32L0X)
 #define EXTI_LINES 30
+#elif defined(CONFIG_SOC_SERIES_STM32L1X)
+#define EXTI_LINES 40
```
Oh yeah... It is 24. Will change it!

Mani-Sadhasivam commented on drivers/i2c/Kconfig.stm32:
```diff
@@ -14,7 +14,7 @@ if I2C_STM32
 
 config I2C_STM32_V1
 	bool "STM32 V1 Driver (F1/F4X)"
```
Ack

Mani-Sadhasivam commented:
@erwango @tagunil Here is the changeset.

* Fixed EXTI lines
* Added IRQ defines for STM32L1X EXTI
* Adopted the GPIO support based on the recently merged STM32 GPIO refactoring PR.

erwango commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -711,6 +730,35 @@ static void __stm32_exti_connect_irqs(struct device *dev)
 		CONFIG_EXTI_STM32_EXTI15_4_IRQ_PRI,
 		__stm32_exti_isr_4_15, DEVICE_GET(exti_stm32),
 		0);
+#elif defined(CONFIG_SOC_SERIES_STM32L1X)
```
Please rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/12463 for factorization

Mani-Sadhasivam replied:
Okay

Mani-Sadhasivam commented on dts/arm/st/l1/stm32l15XXB.dtsi:
```diff
@@ -0,0 +1,22 @@
+ */
+
+#include <mem.h>
+#include <st/l1/stm32l1.dtsi>
```
I don't find any specific difference to add `stm32l151.dtsi` file, that's why I added `stm32l15XXB.dtsi` with common `stm32l1.dtsi`. Can you please help?

erwango replied:
No problem, stm32l151.dtsi could be empty.
Aim is to have same hierarchy as on other series, which proved to scale easily when new SoCs have to be ported.

Mani-Sadhasivam replied:
Okay, cool. Added!

Mani-Sadhasivam commented on soc/arm/st_stm32/stm32l1/soc.h:
```diff
@@ -0,0 +1,60 @@
+#ifndef _STM32L1_SOC_H_
+#define _STM32L1_SOC_H_
+
+#define GPIO_REG_SIZE         0x400
```
Okay

Mani-Sadhasivam commented on drivers/pinmux/stm32/pinmux_stm32l1x.h:
```diff
@@ -0,0 +1,152 @@
+ */
+
+/* Port A */
+#define STM32L1X_PINMUX_FUNC_PA0_PWM2_CH1   \
```
Most of them are copied from L4X. Will cross check this in next iteration.

erwango replied:
ok, txs

Mani-Sadhasivam commented on boards/arm/96b_wistrio/96b_wistrio.yaml:
```diff
@@ -0,0 +1,11 @@
+ram: 16
+flash: 128
+supported:
+  - gpio
```
Okay

Mani-Sadhasivam commented on soc/arm/st_stm32/stm32l1/soc.h:
```diff
@@ -0,0 +1,60 @@
+#include <stm32l1xx_ll_i2c.h>
+#endif
+
+#ifdef CONFIG_SPI_STM32
```
Okay

erwango commented:
@Mani-Sadhasivam , fyi, #12463 has been merged.

Mani-Sadhasivam committed:
drivers: gpio: Add STM32L1X GPIO support

Add GPIO driver support for STM32L1X SoC series.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: interrupt_controller: Add STM32L1X EXTI support

Add EXTI support for STM32L1X SoC series.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam commented:
@erwango I couldn't reproduce the Shippable failure. Is it due to commits reordering by github? Or I'm missing something?

erwango commented on soc/arm/st_stm32/stm32l1/soc.h:
```diff
@@ -0,0 +1,52 @@
+#include <kernel_includes.h>
+
+#include <stm32l1xx_ll_system.h>
+
```
Miss
```
#ifdef CONFIG_EXTI_STM32
#include <stm32l1xx_ll_exti.h>
#endif
```

erwango replied:
@Mani-Sadhasivam, By the way, can I ask you to have a cleaner split of your commits between wrt functionalities addition? I know this requires bit more work on your side, but for instance, it's harder to review when I2C changes are split among multiple commits.

Mani-Sadhasivam committed:
soc: arm: st_stm32: Add STM32L1 SoC series

Add STM32L1 SoC series support with STM32L15XXB as the target
SoC.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
soc: arm: stm32: stm32l1: Add UART support

Add UART support for STM32L1 series SoCs.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

erwango commented on soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb:
```diff
@@ -15,4 +15,26 @@ config NUM_IRQS
 
+if GPIO_STM32
+
+config GPIO_STM32_PORTA
```
GPIOA, B and C are enabled in `st_stm32/common`

Mani-Sadhasivam replied:
Okay

erwango commented on soc/arm/st_stm32/stm32l1/dts_fixup.h:
```diff
@@ -8,4 +8,20 @@
+#define DT_USART_2_IRQ				DT_ST_STM32_USART_40004400_IRQ_0
+#define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
+#define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+
```
Add USART3 since it is provided in dts

Mani-Sadhasivam replied:
Okay

erwango reviewed (changes requested):
Few comments otherwise sounds good.

Mani-Sadhasivam commented on dts/arm/st/l1/stm32l15XXB.dtsi:
```diff
@@ -0,0 +1,22 @@
+/*
```
Done

Mani-Sadhasivam commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -322,6 +322,27 @@ config CLOCK_STM32_PLL_DIVISOR
 
+if SOC_SERIES_STM32L1X
+
+config CLOCK_STM32_PLL_MULTIPLIER
```
Okay

Mani-Sadhasivam commented on soc/arm/st_stm32/stm32l1/dts_fixup.h:
```diff
@@ -78,4 +78,14 @@
 #define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
 #define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
 
+#define DT_I2C_1_BASE_ADDRESS               DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS
```
Done

Mani-Sadhasivam commented on soc/arm/st_stm32/stm32l1/soc.h:
```diff
@@ -47,6 +47,10 @@
 #include <stm32l1xx_ll_exti.h>
 #endif
 
+#ifdef CONFIG_I2C_STM32_V1
```
OKay

Mani-Sadhasivam commented on boards/arm/96b_wistrio/dts_fixup.h:
```diff
@@ -0,0 +1,14 @@
+ * generated data matches the driver definitions.
+ */
+
+#define DT_LIS3DH_DEV_NAME			DT_ST_STM32_I2C_V1_40005400_ST_LIS3DH_32_LABEL
```
Okay

erwango commented on soc/arm/st_stm32/stm32l1/Kconfig.series:
```diff
@@ -0,0 +1,17 @@
+	bool "STM32L1x Series MCU"
+	select CPU_CORTEX_M3
+	select SOC_FAMILY_STM32
+	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
```
Has been renamed into SYS_POWER_LOW_POWER_STATES_SUPPORTED

Mani-Sadhasivam replied:
Done

erwango reviewed (changes requested):
Last update otherwise it should be fine

Mani-Sadhasivam commented:
@erwango I think this PR won't make it for v1.14 LTS?

erwango commented:
> @erwango I think this PR won't make it for v1.14 LTS?

Indeed, it won't. But it will be soon merged once 1.15 window is open.

galak commented:
Some minor things that need fixing up.

frantony commented:
@Mani-Sadhasivam 

Something is wrong with ```_ClearFaults```:

```shell
$ sudo apt-get install -y wget python3 python3-yaml python3-pyelftools make cmake gperf git
$ cd ~
$ wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.10.0/zephyr-sdk-0.10.0-setup.run
$ mkdir zephyr-sdk
$ cd zephyr-sdk/
$ export TMPDIR=/tmp
$ sh ../zephyr-sdk-0.10.0-setup.run -- -d $(pwd)

$ cd ~
$ git clone https://github.com/Mani-Sadhasivam/zephyr
$ cd zephyr
$ git checkout 96b_wistrio
$ git describe
v1.8.99-15325-ga3d76ee3ef
$ export ZEPHYR_TOOLCHAIN_VARIANT=zephyr
$ export ZEPHYR_SDK_INSTALL_DIR=~/zephyr-sdk
$ source zephyr-env.sh
$ cd samples/hello_world
$ rm -rf build && mkdir build && cd build
$ cmake -DBOARD=96b_wistrio ..
$ make

  ...

[ 94%] Building C object zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj
[ 95%] Linking C executable zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       11212 B       128 KB      8.55%
            SRAM:        4116 B        16 KB     25.12%
        IDT_LIST:         120 B         2 KB      5.86%
/home/antony/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: libzephyr.a(soc.c.obj): in function `stm32l1_init':
/home/antony/zephyr/soc/arm/st_stm32/stm32l1/soc.c:35: undefined reference to `_ClearFaults'
collect2: error: ld returned 1 exit status
make[2]: *** [zephyr/CMakeFiles/zephyr_prebuilt.dir/build.make:96: zephyr/zephyr_prebuilt.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:455: zephyr/CMakeFiles/zephyr_prebuilt.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

galak commented:
> Something is wrong with `_ClearFaults`:

it got renamed to `z_clearfaults` if I'm not mistaken.

frantony commented:
@galak 

> > Something is wrong with _ClearFaults:
>
> it got renamed to z_clearfaults if I'm not mistaken.

You are right! Thanks! It got renamed in fd42bf7443cb2f36619be0eab7aa378c30f49da7.

Mani-Sadhasivam commented:
@galak Fixed the issues!

Mani-Sadhasivam committed:
CODEOWNERS: Add codeowner for 96b_wistrio

Add myself as the codeowner for 96Boards Wistrio.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
